Refactor code to improve performance

joshuachp opened the pull request:
Changed the status line function to just get a buffer variable for the
status line and delegate the update of this variable to the lsp_handlers

joshuachp committed:
Refactor code to improve performance

Changed the status line function to just get a buffer variable for the
status line and delegate the update of this variable to the lsp_handlers

wbthomason commented on lua/lsp-status.lua:
```diff
@@ -79,23 +79,22 @@ local function on_attach(client)
-  vim.api.nvim_command('au User LspMessageUpdate redrawstatus!')
-  vim.api.nvim_command('au User LspStatusUpdate redrawstatus!')
-  vim.api.nvim_command('augroup END')
+  vim.api.nvim_command('au User LspDiagnosticsChanged if get(b:, "lsp_status_redraw", 0) | redrawstatus! | end')
+  vim.api.nvim_command('au User LspMessageUpdate if get(b:, "lsp_status_redraw", 0) | redrawstatus! | end')
+  vim.api.nvim_command('au User LspStatusUpdate if get(b:, "lsp_status_redraw", 0) | redrawstatus! | end')
```
Should the check for `lsp_status_redraw` be on all of these? Why not just `LspMessageUpdate`? It seems like the rest are likely to be correctly handled already.

joshuachp replied:
I used the variable to provide extra checks on the redraw function like if the status text is the same we shouldn't redraw. Also, I didn't know if the `autocmd` was used only by the plugin or was exposed to the user. My reasoning was the `event` is issued every time there is an update and the redraw only if needed.

Maybe it could be useful to have a parameter in the `update_lsp_statusline(new_status)` that will check if the function needs to get the new text from diagnostics and messages with `get_lsp_statusline`.

Also, the initial intention was to limit the redraw of the status line every 1/30 seconds since I don't think vim is doing it. I wanted to do it with a check in the `update_lsp_statusline` function that will set the `lsp_status_redraw`. Unfortunately, from my limited understanding of lua, `os.time` function is only in seconds and the  `os.clock` if the cpu time and not the real time, so I didn't add this functionality.

joshuachp replied:
I believe this is why coc.nvim doesn't lag, source code link [coc.nvim/src/model/status.ts](https://github.com/neoclide/coc.nvim/blob/0a565835ea666795fa4b17520cd8f148d074d96a/src/model/status.ts#L14)

wbthomason reviewed (commented):
Thanks for this! I think this is a good direction to take, but there are a few issues with the current implementation. First, we would ideally not call the (relatively expensive) `get_lsp_statusline` function on every message update. I'm playing around with an alternative to this. Second, this seems to break messages with other language servers for me - for example, with Sumneko's Lua LSP, the messages just freeze.

I'll try to fix things up this weekend and ping you for review. Thanks again!

joshuachp commented:
> Thanks for this! I think this is a good direction to take, but there are a few issues with the current implementation. First, we would ideally not call the (relatively expensive) `get_lsp_statusline` function on every message update. I'm playing around with an alternative to this. Second, this seems to break messages with other language servers for me - for example, with Sumneko's Lua LSP, the messages just freeze.
> 
> I'll try to fix things up this weekend and ping you for review. Thanks again!

I'm happy to help, could you elaborate on the issue? I found that the last message on some servers are not updated with ✔️ until you change function and so issue another update.

joshuachp committed:
Implemented a timer base update instead of autocmd

Using the libuv loop to schedule a timer base interval to update the
statusline. This will schedule the tasks to redraw the statusline every
100ms and if the statusline is different.

NOTE: There is an issue with the status text not updating that needs to
be fixed

joshuachp committed:
Fixed timer functionality

Now the timer function work but the last message is still displayed on
the screen.

joshuachp committed:
Cleanup unused variables

joshuachp commented:
I've added the timer to defer the `get_lsp_status` call every 100ms. It uses the vim buffer variables as a global state to know when there is an update. Which probably isn't good but it's working.

Also I noticed the statusline text creates an array and then concatenates all the elements together, which is more work than concatenating each element to a sting.

wbthomason commented:
Thanks again for this. I think the issue I was seeing is what you mention, that the last message is not updated until triggering another event.

> I've added the timer to defer the get_lsp_status call every 100ms. It uses the vim buffer variables as a global state to know when there is an update. Which probably isn't good but it's working.

Thanks, I'll take another look at the design.

> Also I noticed the statusline text creates an array and then concatenates all the elements together, which is more work than concatenating each element to a sting.

Yes, this is more expensive than a fixed, known number of concatenations, but is more convenient and more efficient than concatenating in a loop (ref: https://gitspartv.github.io/LuaJIT-Benchmarks/#test18).

wbthomason commented:
Oh, minor nitpick: can I request that you rename this branch to `refactor/throttle-messaging`? `refactor` is a bit vague, and - while I think I could change the branch name on my end - I think it'll be easier from yours. Thanks!

wbthomason commented:
One suggestion, though I haven't gone through the changes in depth yet: what if, instead of a constantly running timer, we test if the time since last update is within one interval-length of the current time, and issue a `redrawstatus!` if it is not? That would simplify the flow of the code a bit, and potentially save on computation (though probably not beyond the level of a micro-optimization).
